### ISSUE TITLE
Improve invocations retry

### ISSUE THREAD

**Issue by omerb01** [pywren/pywren-ibm-cloud#209]
While computing massive amount of actions, some actions retry to be invoked couple of times and then fail.
I think that this behaviour is problematic in general (particularly in our use case) because If I have massive amount of actions to invoke (~6K in my case) and each one of them takes long time (~100S in my case), all PyWren process will finally fail because some actions will try couple of times till the maximum times have been reached.
I suggest to build a configurable feature (by pywren's config) to retry launch invocations infinity number of times per timestamps. for example lets say: 10sec for the first time, 20sec for the second time and 30 secs for anytime else.

**Comment by omerb01** [comment]
@JosepSampe the idea of a thread pool which I brought in #160 will be better, but its a bit more complicated to deal with. with a thread pool queue we can prioritise invocations by their wait time.

in addition, @gilv suggested to do that dynamic - this means to firstly take a sample of how long each first action takes and then tune each retry by its sampled run time.

what do you think?

**Comment by JosepSampe** [comment]
@omerb01 I agree that the retry mechanism should be improved. In fact some weeks ago I tried to implement another approach similar to what you suggest, but i had some concerns:

Now, when you want to spawn less that 1200 functions, the **invoker** performs all the invocations in 2-3 seconds and unlocks the user code. If you want to spawn much more, the user-code is blocked until the **invoker** finishes. This means that in your case, the user code is blocked like 500 seconds (if you spawn 6k) in the `map()` call. My main concern is that: is this user-friendly? In my case I was thinking about to make the `invoker` as a separate process of the main execution. Then, send the job-description to that process and immediately unlock the user-code. However, this has the main constrain that futures wouldn't be immediately returned to the user. **Invoker** as a separate process could also monitor activations, and spawn more functions as soon as an activation finishes its execution, without the need to set `retry_attemps` and `retry_seconds_list`.

**Comment by omerb01** [comment]
@JosepSampe I think its sounds good. the main thing we need to ensure is that `get_result()` method will continue to be a barrier between done and undone invocations. this means that by passing `get_results()` method, all invocations should be done.

**Comment by omerb01** [comment]
@JosepSampe I think that because of current "retry" mechanism, I experience "read time out" from COS in the first invocation while trying to invoke ~7k actions. I think that your suggestion to move out the invoker as sub-process will prevent thinks like that also.

**Comment by gilv** [comment]
@JosepSampe how complicated to implement it? ;)

**Comment by JosepSampe** [comment]
@gilv work already in progress ;)

**Comment by omerb01** [comment]
@JosepSampe do you have any estimate for this patch? we work on a workflow of thousands actions and its quite hard to follow unexpected exceptions that come from parallelism.

**Comment by JosepSampe** [comment]
@omerb01 Here it is ;) #226 

Due to some performance degradation I found by using only a separate process to invoke all the actions, I finally implemented an hybrid approach that uses the traditional invocation mechanism for the first 1200 actions, and a separate process for the rest of activations and those that fail due to max concurrency.

I think it might be improved, but for now we can already start to test it and share our experiences with it ;)

**Comment by JosepSampe** [comment]
Now with the new invoker we can invoke any number of functions, say 10000,  without using the retrying mechanism with fixed time sleeps. However, we must consider that the performance is different depending of whether we use RabbitMQ or not.

- New invoker based on COS (warm containers):
![invoker_cos](https://user-images.githubusercontent.com/10448186/69732443-eca93700-112b-11ea-978a-efd2d89bb8ad.png)

- New invoker based on RabbitMQ (warm containers):
![invoker_rabbitmq](https://user-images.githubusercontent.com/10448186/69733002-f0898900-112c-11ea-8ffa-8e8cd2d1d416.png)

With the previous tests we can easily see the improvement achieved by using a message broker instead of polling the object storage each X seconds. So for serious experiments I strongly suggest to use RabbitMQ.

**Comment by omerb01** [comment]
@JosepSampe Thanks, tested it a bit and it works fine

**Comment by omerb01** [comment]
@JosepSampe found unexpected behaviour - when `get_result()` fails duo to some error that sent from an action, the new invoker keep launching actions, I think that it should send a signal to the invoker to stop running.

**Comment by JosepSampe** [comment]
Added stop() signal in #232

**Comment by omerb01** [comment]
@JosepSampe it seems that `get_result()` raises "10min timeout" error with the new invoker when 10min are passed regardless of the actual state of the actions.
for example, if we have 500 actions that each one of them takes 8min and we use 250 workers. so in the first step we launch 250 actions in 8min and then launch again 250 actions. after another 2min we get "10min timeout" error.
can you check it please?

**Comment by JosepSampe** [comment]
@omerb01 Yep, default waiting timeout is 10 minutes. I'm not sure if we can dynamically predict the timeout for each use cae. In any case, you can change this timeout in the get_result() method, for example get_result(timeout=800)

**Comment by omerb01** [comment]
@JosepSampe is it possible to implement a timer for each job?

**Comment by JosepSampe** [comment]
@omerb01 What do you mean by "implement a timer for each job"?

**Comment by gilv** [comment]
@omerb01 i also a bit confused...if you submit a job and it get to 10 min then you get "10 min timeout" error...what is the problem you observe?

**Comment by omerb01** [comment]
@gilv @JosepSampe example code:
```python
import pywren_ibm_cloud as pywren
import time

def my_func(args):
    time.sleep(540)
    return "Hello World!"

pw = pywren.ibm_cf_executor(workers=1)
pw.map(my_func, range(2))
print(pw.get_result())
```
this code raises "timeout" error but it should run succefully...
the problem is that the second action gets only 1min run time instead of 10min available from CF.

I think that it's not a good approach to set a global timeout for `get_result()` method. I suggest to set a 10min timer per each action (it should be the same timeout as defined in CF) and then each action will have this time to run.
I think it's better because in this way, each action will have the same time to run.

**Comment by JosepSampe** [comment]
@omerb01 Each action already has a timeout of 10 minutes defined in config by default..Timeout stated in get_result is completely diferent, it is a timeout raised on the client side and used to unlock the get_result method for not being waiting forever for a function result

**Comment by omerb01** [comment]
@JosepSampe so why do we need this timeout? I think that while we have "live" actions, we should wait to its results by default

**Comment by JosepSampe** [comment]
@omerb01 The problem is, for example, an OOM error. Sometimes functions are not executed properly, so there is no way to know if a function is still running or if a function failed its execution. With this timeout you can unlock the local execution and not being waiting forever for a function result that does not exist. 

Maybe the normal usage of PyWren is to spawn <= 1000 functions, so this timeout is aligned with the timeout of the functions. In the other cases, if you know that you have to run for example 20 steps of 200 functions that take more or less 5 minutes each one to finish, you can configure this parameter and increase the timeout.

Other reason is that maybe sometimes you don't need all the results from the functions and you can assume that some function activation failed. So with this timeout you can continue the local execution without having to cancell the whole experiment an reexecute the entire job (or jobs) again.

**Comment by omerb01** [comment]
@JosepSampe OOM is a problem that we haven't solved yet, PyWren should detect actions that got OOM errors. I found it extremely hard to debug and develop a workload contains thousands of actions that I can't estimate how long each one will take because it depends on many things and it is individual for each dataset. I expect that `get_result()` won't limit my workload and if a single action gets a "timeout" error, `get_result()` will stop and raise an error.

**Comment by omerb01** [comment]
@JosepSampe does `get_result()` detect actions that got "timeout" errors? (regardless the global timeout parameter)

**Comment by JosepSampe** [comment]
@omerb01 Yep, it's true that sometimes we can't predict the global timeout. I only exposed the reasons of that timeout based on my experience ;)  I think we can change the behavior of this timeout and set it to None by default, and if someone wants to limit the local waiting time, he can set the desired global timeout.

> @JosepSampe does `get_result()` detect actions that got "timeout" errors? (regardless the global timeout parameter)
Yep

**Comment by omerb01** [comment]
@JosepSampe in addition, regarding OOM - I think that we can set an assumption on actions that run more than 10min. `get_result()` shouldn't wait forever, it should detect that 10min have passed and raise "fatal error" or something similar (since it's not timeout)

**Comment by gilv** [comment]
@JosepSampe thanks a lot for the pull request. 
i am a bit confused with this thread... @omerb01 claims if some invocation not killed, then get_result() will stuck for ever? is this what you also observe?
Also - each worker has timeout which is global per all invocation used inside? so it's not timeout per invocation, but per worker?

**Comment by JosepSampe** [comment]
@gilv 

> @JosepSampe thanks a lot for the pull request.
> i am a bit confused with this thread... @omerb01 claims if some invocation not killed, then get_result() will stuck for ever? is this what you also observe?

get_result() method has an available parameter to set a global timeout for waiting for the results, for example `get_result(timeout=1000)`. This means that your local pywren code will be blocked during at most 1000 seconds. If during these 1000 seconds get_result() is unable to download all the results, it will raise a timeout and continue local execution.

> Also - each worker has timeout which is global per all invocation used inside? so it's not timeout per invocation, but per worker?

No, each activation has its own 10 minutes as always, we don't changed this behavior ;) There is no timeout per worker.

**Comment by omerb01** [comment]
@JosepSampe experienced a new unexpected behavior when I use PyWren with the new invoker.
when I use Jupyter with one single PyWren executor which initialized in a separated cell, and then I start invocations to the max workers and stopped the executor by force, the next time I'll run `map()` and `get_result()`, the new invoker won't launch any invocation cause it will wait for invocations to be done (but its actually a new execution).
I assume the new invoker should be initialised after each `map()` execution instead of initialise it in `ibm_cf_executor()`

**Comment by gilv** [comment]
@JosepSampe if i don't set timeout and some invocation killed completely, then get_result() will stuck forever?

**Comment by JosepSampe** [comment]
@gilv Yes, until you press ctrl+c

**Comment by omerb01** [comment]
@JosepSampe I'm sorry if this thread turns confusing. regarding my last comment, I found that with the current invoker, it's not possible to use one executor for multiple `map()` executions using Jupyter - I experience weird behavior when `get_result()` waits in vain. is it possible to fix?

**Comment by JosepSampe** [comment]
@omerb01 New invoker is initiated in the executor, this fact allows call_async(), map() and map_reduce() to be async methods, i.e, they submit the job to the invoker and immediately return the control to the user code. 

To solve the issue you found we can revert previous pywren approach where the invoker is started in each map() call_async(), map_reduce(). However, with this approach the map() call_async(), map_reduce() methods will be synchronous, i.e. the method will stuck user code until all the activations are invoked. This means that, for example, if your functions takes 10 minutes to finish and you want to invoke 3000 activations, the map() method will stuck the user code during 20+ minutes.

**Comment by JosepSampe** [comment]
@omerb01 In your previous comment you say: "..., and then I start invocations to the max workers and stopped the executor by force..." What do you mean by "stopped by force"? Do you refer to press the "stop (interrupt kernel)" button of the jupyter interface?

**Comment by JosepSampe** [comment]
@omerb01 Added fix in #245. Try if you get expected behavior when merged
